src/System.Private.CoreLib/shared/System/MemoryExtensions.Fast.cs

```diff
@@ -400,6 +400,53 @@ public static Span<T> AsSpan<T>(this T[] array, int start)
             return new Span<T>(ref Unsafe.Add(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), start), array.Length - start);
         }
 
+        /// <summary>
+        /// Creates a new span over the portion of the target array.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Span<T> AsSpan<T>(this T[] array, Index startIndex)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            if (default(T) == null && array.GetType() != typeof(T[]))
+                ThrowHelper.ThrowArrayTypeMismatchException();
```

Comment:
Now I understand why Jan was asking about a mismatch exception. Seems odd in some cases to throw and in others to use the actual type.

Reply:
Yup. I used you implementation which seams work perfectly.